src/data/sequence.rs

```diff
@@ -0,0 +1,788 @@
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+// #![allow(dead_code)] // for the draft PR only
+
+use crate::authorization::access_control::{
+    AccessListTrait, AccessType, PrivateAccessList, PrivateUserAccess, PublicAccessList,
+    PublicUserAccess,
+};
+use crate::shared_data::{
+    to_absolute_range, to_absolute_version, Address, ExpectedVersions, Kind, NonSentried, Owner,
+    Sentried, User, Value, Version,
+};
+use crate::{Error, PublicKey, Result, XorName};
+use serde::{Deserialize, Serialize};
+use std::fmt::{self, Debug, Formatter};
+
+pub type PublicSentriedSequence = SequenceBase<PublicAccessList, Sentried>;
+pub type PublicSequence = SequenceBase<PublicAccessList, NonSentried>;
+pub type PrivateSentriedSequence = SequenceBase<PrivateAccessList, Sentried>;
+pub type PrivateSequence = SequenceBase<PrivateAccessList, NonSentried>;
+pub type Values = Vec<Value>;
+
+/// A representation of data at some version.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Default, Debug)]
+pub struct DataEntry {
+    /// Data version
+    pub version: u64,
+    /// Data value
+    pub value: Vec<u8>,
+}
+
+impl DataEntry {
+    /// Returns a new instance of a data entry.
+    pub fn new(version: u64, value: Vec<u8>) -> Self {
+        Self { version, value }
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash)]
+pub struct SequenceBase<P, S> {
+    address: Address,
+    data: Values,
+    access_list: Vec<P>,
+    // This is the history of owners, with each entry representing an owner.  Each single owner
+    // could represent an individual user, or a group of users, depending on the `PublicKey` type.
+    owners: Vec<Owner>,
+    _flavour: S,
+}
+
+/// Common methods for all `Sequence` flavours.
+impl<C, S> SequenceBase<C, S>
+where
+    C: AccessListTrait,
+    S: Copy,
+{
+    /// Returns true if the provided access type is allowed for the specific user (identified y their public key).
+    pub fn is_allowed(&self, user: PublicKey, access: AccessType) -> bool {
+        if let Some(owner) = self.owner_at(Version::FromEnd(1)) {
+            if owner.public_key == user {
+                return true;
+            }
+        }
+        match self.access_list_at(Version::FromEnd(1)) {
+            Some(list) => list.is_allowed(&user, access),
+            None => false,
+        }
+    }
+
+    /// Return the address of this Sequence.
+    pub fn address(&self) -> &Address {
+        &self.address
+    }
+
+    /// Return the name of this Sequence.
+    pub fn name(&self) -> &XorName {
+        self.address.name()
+    }
+
+    /// Return the type tag of this Sequence.
+    pub fn tag(&self) -> u64 {
+        self.address.tag()
+    }
+
+    /// Returns true if the user is the current owner.
+    pub fn is_owner(&self, user: PublicKey) -> bool {
+        match self.owner_at(Version::FromEnd(1)) {
```

Comment:
so only one owner allowed?

Reply:
BLS key gives multiple owner under 1 key. This is just copied from previous implementation, nothing changed, except from wording. It was called "owners" before, buit validation was always made against current entry. The vector of owner was in fact the _history_ of owners, but it was not expressed in the naming.

Reply:
but the type of user is `PublicKey`, which normally means `a public key for just one user`
If we do want to use BLS key (or plan to do so later on), maybe change the type to something like `OwnershipKey`, and currently define `type OwnershipKey = PublicKey`, and later on switch it to BLSKey ?

Reply:
Yep, I think that's a good idea.

Reply:
Saw now that the question of PublicKey is moot, since it already is defined like this:

```
pub enum PublicKey {
    Ed25519(ed25519_dalek::PublicKey),
    Bls(threshold_crypto::PublicKey),
    BlsShare(threshold_crypto::PublicKeyShare),
}
```

Reply:
I am wondering that whether the network can allow different type of `PublicKey` co-exists within the network. (i.e. some node using `Ed25519` as PublicKey, meanwhile others using `Bls`).
Also, not sure about the diff between `Bls` and `BlsShare`, can they co-exist?
So, maybe making `PublicKey` as enum is un-necessary?

Reply:
I'd also like to get a clarification about this, or do we need to use spefici type of keys in different areas of the network, e.g. a SafeKey (CoinBalance) always is on a BLS PK, so all code related to managing this type of data uses a PublicKey::Bls/BlsShare, etc.

Moreover, why is it we need the Ed25519? can't we use BLS for everything?